### PR TITLE
feat: delist and change price 

### PR DESCRIPTION
**Pull request by AlefrankM**

### CODE REVIEW

File: src/features/marketplace/process/clients/delist-item-client.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import StoreDelistItemCommand from '../commands/delist-item/store-delist-item-command';
+import IClient from '../interfaces/client.interface';
+import DelistItemInvoker from '../invokers/delist-item-invoker';
+import MarketplaceState from '../types/marketplace-state';
+
+export default class DelistItemClient implements IClient {
+  private readonly invoker: DelistItemInvoker;
+
+  constructor(state: MarketplaceState) {
+    console.log(state);
```
Comment (GimerStackIt) [inline src/features/marketplace/process/clients/delist-item-client.ts:14]: Remove Console.log

**Review by juliswer — APPROVED** [review]
Looks good.
Just a comment: please try to make shorter PRs, otherwise will be getting hard finding a possible vulnerability, bug or whatever to be fixed before merging main.
If you have many changes, try sending daily PRs, so is easier check you progress.

**Review by GimerStackIt — APPROVED** [review]

File: src/components/item-details/bid-tab/delisting-tab-content.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+export const DelistingTabContent = ({ item }: Props) => {
+  useMetamask();
+  const { isLoading } = useAppSelector((state) => state.marketplace);
+  const [show, setShow] = useState(false);
+  const dispatch = useAppDispatch();
+
+  const onSubmit = async () => {
+    dispatch(
```
Comment (mmiranda-stackit) [inline src/components/item-details/bid-tab/delisting-tab-content.tsx:22]: Maybe it is good idea to include the withProtection action in here, so we double check that the action should be performed by an authenticated user.

File: src/features/marketplace/process/commands/change-price-item/store-change-price-item-command.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  private readonly itemService: ItemService;
+
+  constructor(_itemService: ItemService) {
+    this.itemService = _itemService;
+  }
+
+  async execute(state: MarketplaceState): Promise<MarketplaceState> {
+    if (!state.itemId) return { ...state, error: MarketplaceError.RequiredItemId };
+    if (!state.price) return { ...state, error: MarketplaceError.RequiredPrice };
+    if (!state.marketplaceEvent) return { ...state, error: MarketplaceError.RequiredListEvent };
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/change-price-item/store-change-price-item-command.ts:16]: Why do you need the marketplaceEvent here? And if you happen to need it, you are returning the wrong error.

File: src/features/marketplace/process/commands/change-price-item/store-change-price-item-command.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      state.item = await this.itemService.list(
+        state.itemId,
+        state.price,
+        state.listId,
+        state.ownerAddress
+      );
+    } catch (ex) {
+      // TODO: Handle exceptions properly
+      console.log('ex|StoreChangePriceItemCommand', ex);
+      return { ...state, error: MarketplaceError.StoreListItemFailure };
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/change-price-item/store-change-price-item-command.ts:28]: You should throw the proper error here, StoreChangePriceFailure or something like that

File: src/features/marketplace/process/commands/delist-item/change-status-item-command.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  private readonly marketplaceService: IMarketplaceService;
+
+  constructor(_marketplaceService: IMarketplaceService) {
+    this.marketplaceService = _marketplaceService;
+  }
+
+  async execute(state: MarketplaceState): Promise<MarketplaceState> {
+    if (!state.listId) return { ...state, error: MarketplaceError.RequiredListId };
+    if (!state.itemId) return { ...state, error: MarketplaceError.RequiredTokenId };
+    if (!state.ownerAddress) return { ...state, error: MarketplaceError.RequiredOwnerAddress };
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/delist-item/change-status-item-command.ts:17]: This ownerAddress check is not needed

File: src/features/marketplace/process/commands/delist-item/store-delist-item-command.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export default class StoreDelistItemCommand implements ICommand<MarketplaceState> {
+  private readonly itemService: ItemService;
+
+  constructor(_itemService: ItemService) {
+    this.itemService = _itemService;
+  }
+
+  async execute(state: MarketplaceState): Promise<MarketplaceState> {
+    if (!state.itemId) return { ...state, error: MarketplaceError.RequiredItemId };
+    if (!state.marketplaceEvent) return { ...state, error: MarketplaceError.RequiredListEvent };
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/delist-item/store-delist-item-command.ts:15]: You are missing the ownerAddress check. The marketplaceEvent is not required here

File: src/features/marketplace/process/commands/delist-item/store-delist-item-command.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    try {
+      state.item = await this.itemService.delist(state.itemId, state.ownerAddress);
+    } catch (ex) {
+      // TODO: Handle exceptions properly
+      console.log('ex|StoreDelistItemCommand', ex);
+      return { ...state, error: MarketplaceError.StoreListItemFailure };
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/delist-item/store-delist-item-command.ts:22]: The error is wrong. You should return the StoreDelistItemFailure instead

File: src/features/marketplace/process/invokers/list-item-invoker.ts [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
     private onListItemCommand: ICommand<MarketplaceState>,
-    private onStoreListItemCommand: ICommand<MarketplaceState>
+    private onStoreDelistItemCommand: ICommand<MarketplaceState>
   ) {}
 
   async execute() {
     this.state = await this.onListItemCommand.execute(this.state);
 
-    if (!this.state.error) this.state = await this.onStoreListItemCommand.execute(this.state);
+    if (!this.state.error) this.state = await this.onStoreDelistItemCommand.execute(this.state);
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/invokers/list-item-invoker.ts:14]: How come you are changing list by delist in the list item invoker?

File: src/features/marketplace/store/marketplace.actions.ts [hunk 9]
```diff
@@ -93,6 +135,51 @@ export const buyItem = createAsyncThunk(
...
+        itemId,
+        listId,
+        status: ItemStatus.Listed,
+      } as MarketplaceState;
+
+      const processor = new MarketplaceClientProcessor();
+      const listed = await processor.execute(marketplaceState);
+
+      dispatch(
+        openToastSuccess('The item has been successfully changed price on the marketplace.')
```
Comment (mmiranda-stackit) [inline src/features/marketplace/store/marketplace.actions.ts:170]: Change the label to something like 'The item price has been successfully changed"

**Review by mmiranda-stackit — CHANGES_REQUESTED** [review]
See my comments. I didn't see any unit test here.

File: src/components/item-details/bid-tab/delisting-tab-content.tsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        </Modal.Header>
+        <Modal.Body>
+          <p className="text-center">You are about to delist your NFT from Marketplace</p>
+          <div className="placebid-form-box">
+            <div className="bit-continue-button">
+              <Button
+                size="medium"
+                fullwidth
+                onClick={onSubmit}
+                className={isLoading ? 'disabled' : ''}
```
[the remark this answers is not in the document]
Comment (AlefrankM) [reply src/components/item-details/bid-tab/delisting-tab-content.tsx:81]: Yeas, you can't click it

File: src/features/marketplace/process/commands/change-price-item/change-price-item-command.ts [hunk 11]
```diff
@@ -36,7 +36,10 @@ export default class ChangePriceItemCommand implements ICommand<MarketplaceState
       state.marketplaceEvent = eventFound;
       state.mintEventName = ContractEvent.LogChangeStatus;
     } catch (ex) {
-      return rejectWithMetamask(ex, () => ({ ...state, error: MarketplaceError.ListItemFailure }));
+      return rejectWithMetamask(ex, () => ({
+        ...state,
+        error: MarketplaceError.ChangeStatusItemFailure,
```
Comment (mmiranda-stackit) [inline src/features/marketplace/process/commands/change-price-item/change-price-item-command.ts:41]: ChangeItemPriceFailure?

**Review by mmiranda-stackit — APPROVED** [review]
nicely done

**Pull request merged**